File: tests/parity/python/pipeline.py
"""Python side of the cross-language parity pipeline.

Walks a corpus of Markdown files, splits each with the shared Rust recursive
splitter, and writes one deterministic ``*.chunks`` file per source into a
``localfs`` DirTarget. The output-file bytes are constructed identically to the
TypeScript driver (``tests/parity/ts/pipeline.ts``) so the two DirTargets can be
compared byte-for-byte.

Also emits a metrics JSON capturing the per-component ``UpdateStats`` counters,
which the runner uses to assert incremental behaviour (edit one file, re-run,
expect exactly one component to reprocess and the rest to stay unchanged).
"""

from __future__ import annotations

import argparse
import asyncio
import json
import pathlib

import grepify as coco
from grepify.connectors import localfs
from grepify.resources.file import FileLike, PatternFilePathMatcher

# Chunking configuration shared with the TS driver. Keep in sync with
# tests/parity/ts/pipeline.ts.
CHUNK_SIZE = 200
CHUNK_OVERLAP = 40
LANGUAGE = "markdown"

_RECORD_SEP = b"\x1e"


def _render_chunks(relative_path: str, source_bytes: bytes, chunks: list) -> bytes:
    """Build the canonical, cross-language output bytes for one source file.

    Format (identical in Python and TypeScript):
        <relative_path>\n
        <n_chunks>\n
        for each chunk:
            "<sb> <eb> <sc> <ec> <sl> <scol> <el> <ecol>\n"
            <chunk bytes sliced from source by byte offset>
            <0x1e record separator>
    """
    out = bytearray()
    out += relative_path.encode("utf-8") + b"\n"
    out += str(len(chunks)).encode("utf-8") + b"\n"
    for c in chunks:
        header = (
            f"{c.start.byte_offset} {c.end.byte_offset} "
            f"{c.start.char_offset} {c.end.char_offset} "
            f"{c.start.line} {c.start.column} "
            f"{c.end.line} {c.end.column}\n"
        )
        out += header.encode("utf-8")
        out += source_bytes[c.start.byte_offset : c.end.byte_offset]
        out += _RECORD_SEP
    return bytes(out)


_splitter = None


def _get_splitter():
    global _splitter
    if _splitter is None:
        from grepify.ops.text import RecursiveSplitter

        _splitter = RecursiveSplitter()
    return _splitter


@coco.fn(memo=True)
async def process_file(item: tuple[str, FileLike], target: localfs.DirTarget) -> None:
    # `item` is (walk-relative path, file). The relative path (forward slashes,
    # relative to the walk root) is used for the output name and the content
    # header so it matches the TS driver's `FileEntry.relativePath`.
    relpath, file = item
    text = await file.read_text()
    source_bytes = text.encode("utf-8")
    chunks = _get_splitter().split(
        text,
        chunk_size=CHUNK_SIZE,
        chunk_overlap=CHUNK_OVERLAP,
        language=LANGUAGE,
    )
    outname = relpath.replace("/", "__") + ".chunks"
    target.declare_file(
        filename=outname,
        content=_render_chunks(relpath, source_bytes, chunks),
    )


@coco.fn
async def app_main(sourcedir: pathlib.Path, outdir: pathlib.Path) -> None:
    target = await coco.use_mount(localfs.declare_dir_target, outdir)
    files = localfs.walk_dir(
        sourcedir,
        recursive=True,
        path_matcher=PatternFilePathMatcher(included_patterns=["**/*.md"]),
    )
    items = [(key, (key, file)) async for key, file in files.items()]
    handle = await coco.mount_each(
        coco.component_subpath("process"), process_file, items, target
    )
    await handle.ready()


def _stats_to_dict(stats: coco.UpdateStats | None) -> dict:
    if stats is None:
        return {"total": {}, "by_component": {}}

    def group(g) -> dict:
        return {
            "num_execution_starts": g.num_execution_starts,
            "num_unchanged": g.num_unchanged,
            "num_adds": g.num_adds,
            "num_deletes": g.num_deletes,
            "num_reprocesses": g.num_reprocesses,
            "num_errors": g.num_errors,
            "num_processed": g.num_processed,
        }

    return {
        "total": group(stats.total),
        "by_component": {name: group(g) for name, g in stats.by_component.items()},
    }


async def _run(args: argparse.Namespace) -> None:
    env = coco.Environment(coco.Settings.from_env(db_path=args.db))
    app = coco.App(
        coco.AppConfig(name="parity_pipeline", environment=env),
        app_main,
        sourcedir=args.source,
        outdir=args.out,
    )
    handle = app.update()
    await handle.result()
    stats = handle.stats()

    metrics = {
        "host": "python",
        "out_dir": str(args.out),
        "db_path": str(args.db),
        "stats": _stats_to_dict(stats),
    }
    args.metrics.parent.mkdir(parents=True, exist_ok=True)
    args.metrics.write_text(json.dumps(metrics, indent=2), encoding="utf-8")


def main() -> None:
    parser = argparse.ArgumentParser(description="Python parity pipeline")
    parser.add_argument("--source", type=pathlib.Path, required=True)
    parser.add_argument("--out", type=pathlib.Path, required=True)
    parser.add_argument("--db", type=pathlib.Path, required=True)
    parser.add_argument("--metrics", type=pathlib.Path, required=True)
    asyncio.run(_run(parser.parse_args()))


if __name__ == "__main__":
    main()

File: tests/parity/ts/pipeline.ts
// TypeScript side of the cross-language parity pipeline.
//
// Mirror of tests/parity/python/pipeline.py: walk a corpus of Markdown files,
// split each with the shared Rust recursive splitter, and write one
// deterministic `*.chunks` file per source into a localfs DirTarget. The
// output bytes are constructed identically to the Python driver so the two
// DirTargets compare byte-for-byte.
//
// Run with tsx (see package.json). We import the SDK by relative path to
// packages/node/index.ts because the published `@grepify/node` entry
// (index.js) is only generated at publish time; the native `.node` artifact and
// the TS sources are what exist in-repo today.

import { mkdirSync, writeFileSync } from 'node:fs'
import { dirname } from 'node:path'
import {
  App,
  type Chunk,
  fn,
  localfs,
  mountEach,
  splitTextRecursive,
  walkDir,
} from '../../../packages/node/index.ts'

// Chunking configuration shared with the Python driver. Keep in sync with
// tests/parity/python/pipeline.py.
const CHUNK_SIZE = 200
const CHUNK_OVERLAP = 40
const LANGUAGE = 'markdown'

const RECORD_SEP = Buffer.from([0x1e])

interface Args {
  source: string
  out: string
  db: string
  metrics: string
}

function parseArgs(): Args {
  const argv = process.argv.slice(2)
  const map: Record<string, string> = {}
  for (let i = 0; i < argv.length; i += 2) {
    map[argv[i].replace(/^--/, '')] = argv[i + 1]
  }
  return { source: map.source, out: map.out, db: map.db, metrics: map.metrics }
}

/** Build the canonical, cross-language output bytes for one source file. */
function renderChunks(relativePath: string, sourceBytes: Buffer, chunks: Chunk[]): Buffer {
  const parts: Buffer[] = []
  parts.push(Buffer.from(`${relativePath}\n`, 'utf8'))
  parts.push(Buffer.from(`${chunks.length}\n`, 'utf8'))
  for (const c of chunks) {
    const header =
      `${c.startByte} ${c.endByte} ` +
      `${c.startCharOffset} ${c.endCharOffset} ` +
      `${c.startLine} ${c.startColumn} ` +
      `${c.endLine} ${c.endColumn}\n`
    parts.push(Buffer.from(header, 'utf8'))
    parts.push(sourceBytes.subarray(c.startByte, c.endByte))
    parts.push(RECORD_SEP)
  }
  return Buffer.concat(parts)
}

interface FileItem {
  rel: string
  text: string
}

async function main(): Promise<void> {
  const args = parseArgs()

  const app = await App.open('parity_pipeline', { dbPath: args.db })

  // The processor closes over `target` (a DirTarget is not serializable, so it
  // cannot be a mountEach extra-arg without disabling memoization). The memo key
  // is derived from the fn logic hash + the item value, so a content change on
  // one file invalidates only that child.
  const appMain = async (sourceDir: string, outDir: string): Promise<void> => {
    // Create the target in the root component scope, then use it from the
    // `process/*` children via closure (mirrors packages/node/test/target.test.ts).
    const target = localfs.mountDirTarget(outDir)

    const files = walkDir(sourceDir, {
      recursive: true,
      includedPatterns: ['**/*.md'],
    })
    const items: [string, FileItem][] = files.map((f) => [
      f.relativePath,
      { rel: f.relativePath, text: f.contentStr() },
    ])

    const processFile = fn(
      async (item: FileItem): Promise<void> => {
        const sourceBytes = Buffer.from(item.text, 'utf8')
        const chunks = splitTextRecursive(item.text, {
          chunkSize: CHUNK_SIZE,
          chunkOverlap: CHUNK_OVERLAP,
          language: LANGUAGE,
        })
        const outName = item.rel.replaceAll('/', '__') + '.chunks'
        target.declareFile(outName, renderChunks(item.rel, sourceBytes, chunks))
      },
      { memo: true },
    )

    const handle = await mountEach('process', processFile, items)
    await handle.ready()
  }

  const updateHandle = app.startUpdate(appMain, { args: [args.source, args.out] })
  // Drive to completion with `changed()` (which restores the handle each poll)
  // rather than `result()` (which consumes it), so `stats()` stays readable.
  while (await updateHandle.changed()) {
    /* keep advancing until the update terminates */
  }
  const stats = updateHandle.stats()

  const metrics = {
    host: 'typescript',
    out_dir: args.out,
    db_path: args.db,
    stats: {
      ready: stats.ready,
      by_component: Object.fromEntries(
        Object.entries(stats.byComponent).map(([k, g]) => [
          k,
          {
            num_execution_starts: g.numExecutionStarts,
            num_unchanged: g.numUnchanged,
            num_adds: g.numAdds,
            num_deletes: g.numDeletes,
            num_reprocesses: g.numReprocesses,
            num_errors: g.numErrors,
            num_processed:
              g.numUnchanged + g.numAdds + g.numDeletes + g.numReprocesses,
          },
        ]),
      ),
    },
  }

  mkdirSync(dirname(args.metrics), { recursive: true })
  writeFileSync(args.metrics, JSON.stringify(metrics, null, 2), 'utf8')
}

main().then(
  () => process.exit(0),
  (err) => {
    console.error(err)
    process.exit(1)
  },
)
